Feature pagination

aleksejs-fomins opened the pull request:
* Full implementation of BFabric2 interface.
* Pagination enabled through read, read_multi, delete and delete_multi procedures
* Some integration and unit tests
* ResultContainer is adapted to produce responses in list_dict format, such that parsed responses from ZEEP and SUDS are as compatible as possible.

leoschwarz reviewed (changes requested):
Thanks again for your work on this.

aleksejs-fomins commented on bfabric/src/math_helper.py:
```diff
@@ -0,0 +1,20 @@
+    print(
+        div_int_ceil(120, 100),
+        div_int_ceil(200, 100),
+        div_int_ceil(245, 100)
```
The major reason for this function is that `math.ceil(n/d)` applies a ceiling function to a floating-point number. There is a risk that ceil(2/4) = 3 instead of 2 because of roundoff errors. I'm happy to also add the test for large numbers, but that was not the reason to introduce it.

leoschwarz replied:
I'm ok with the function as is, because it's correct and general. But I'm saying it probably is not necessary for our current use case, because, assuming IEEE 754 doubles:

- the concern would apply to cases where the division remainder is 0, because all others are supposed to be rounded-up by `math.ceil` anyhow (ignoring negative numbers for simplicity)
- integer numbers can be [represented exactly](https://en.wikipedia.org/wiki/Double-precision_floating-point_format#Precision_limitations_on_integer_values) as float from -2^53, +2^53 
- (basic operations like `+,-,/,*,sqrt`, i.e. not necessarily `sin(x)` or `pow(a,b)`) floating point operations will be rounded to the nearest exactly representable value (IEEE 754 section 4 intro, section 5)

If you use numbers above the range which is representable correctly, then you might run into problems, which your code solves and the case above would test for.

aleksejs-fomins replied:
I see, thanks for clarifying. I was not aware roundoff issue was only there for large numbers.

aleksejs-fomins commented on bfabric/src/result_container.py:
```diff
@@ -9,18 +17,88 @@ class BfabricResultType(Enum):
+    def __len__(self):
+        return len(self.results)
+
+    def append(self, other: ResultContainer) -> None:
```
I like the suggestion. I really don't understand why is it offering me to commit this by clicking commit suggestion. This is not a full code refactor, and will only change the name of the function, without changing the name in the places where the function is used. This would create a non-functioning code 0_0

leoschwarz replied:
It is, because I created the suggestion to illustrate the point, but you are right that committing just that would not work... I should have added `requires further changes` or something like that to be clear. :)

aleksejs-fomins commented on bfabric/tests/unit/test_results_container.py:
```diff
@@ -0,0 +1,46 @@
+        self.assertEqual(str(self.c1), "[1, 2, 3]")
+        self.assertEqual(str(self.c2), "[4, 5]")
+
+        self.assertEqual(self.c1.__repr__(), "[1, 2, 3]")
+        self.assertEqual(self.c2.__repr__(), "[4, 5]")
```
I did not know you could do that :)

aleksejs-fomins commented on bfabric/bfabric2.py:
```diff
@@ -60,37 +59,172 @@ class Bfabric(object):
+        for page_vals in page_iter(multi_query_vals):
+            obj_exteded[multi_query_key] = page_vals
+
+            # TODO: Test what happens if there are multiple responses to each of the individual queries.
```
I forgot I wanted to do this. I will make a separate pull request later where I will test this and provide corrections for this case specifically, as well as document it

aleksejs-fomins commented:
I have addressed all of your comments. However, all of a sudden, another test in test_bfabric2_read.py no longer passes for Zeep, even though it did on Friday. I contacted Can about this, will likely make another small commit to this branch before we merge, to make sure all tests pass
